### PR TITLE
Distro bootloader specification alternate version

### LINKED ISSUE #24: Implement the distro bootloader specification
I wonder if it's possible to implement the distro bootloader specification ?
(see also doc/README.distro in U-Boot source tree).

Using the bootloader specification, it should be possible to keep the boot menu to switch between Android  and GNU/Linux but the latter will use extlinux to load the kernel, dtb and initramfs along with bootargs, all from external mmc.

Is this could be possible using the current U-Boot tree ? Or is there any constraints because of the chainload that does not permit to load any kernel from external mmc ?

 On a side note, I was able to boot to a Fedora 30 (rawhide) userspace. But I need to revert the removed chosen bootargs node as I don't see where I could have specified the bootargs elsewhere.

Thx for this impressive work.

### LINKED ISSUE #26: Finding a good value for ramdisk address
Using a ramdisk is mandatory to enable support for sysboot (extlinux). But I'm unable to find a good address. Last attempt is with:

ramdisk_addr_r=0x40000000\0"

But then I have the following message as shown by
https://pasteboard.co/I0Pasyy.jpg
Failed to allocate 0x35607db bytes bellow 0x35000000.
ramdisk - allocation error

All components are loaded as appropriate and ramdisk size is 55969755 bytes.
I will try to reduce the size of the ramdisk and try to reproduce. (but default fedora ramdisk will likely be that huge).

### PR DESCRIPTION
**Pull request by xc-racer99**
As a competitor to  #25  :)

This implements the distro bootloader specification and also supports Android booting.  The relevant SD Card/MMC entries show per device.  It is also integrates the sysboot menu with the bootmenu of u-boot so we can use the volume keys to select which of the entries to use.

It has the added advantage from the old way of allowing the order of the bootmenu entries to be more sane (ie update u-boot from SD card at the bottom) as they are dynamically created on boot.  It also changes things in that it's less picky about which buttons are pressed to enter the menu/recovery.

I've tested both Android boot and postmarketOS with an extlinux.conf file (not upstream yet, just locally Edit: Now available at https://gitlab.com/xc-racer99/pmaports/commits/fascinate4g)

In terms of ramdisk_addr_r, I've used 0x40000000 as this will give us the absolute most.  As per my comment at https://github.com/xc-racer99/u-boot-galaxys4g/issues/26#issuecomment-462879183 and why it appears that only pxeboot was using the the ramdisk_addr_r, it's because the syslinux/extlinux stuff all goes through the PXE boot method.

@kwizart As to why the ramdisk_addr_r wasn't working for you, are you sure that you updated the environment properly?  If you installed the new version via the bootmenu, then you would still have had the old value as CONFIG_EXTRA_ENV_SETTINGS do not overwrite pre-existing values.  If you flashed via heimdall/Odin then it would have wiped out the saved env and you would have had the new value.  In any event, we won't get more space anywhere else.

Closes #24
Closes #26

### CODE REVIEW

**Comment by PabloPL** [comment]
Looks Good to me, nice cleanup in board code. I though some time ago about the same thing - using bootmenu entries for pxe/sysboot.

**Comment by PabloPL** [comment]
About postmarketos - i wonder if it would boot using old kernel also (to use gpu accel).

**Comment by xc-racer99** [comment]
> About postmarketos - i wonder if it would boot using old kernel also (to use gpu accel).

That's my thought too.  I must have messed something up at the moment though as I only get the "Uncompressing Linux...  Done booting the kernel" but nothing else then, even with the FIQ disabled.  I'll try again today and see if I can figure anything out.

**Comment by kwizart** [comment]
> @kwizart As to why the ramdisk_addr_r wasn't working for you, are you sure that you updated the environment properly?...

I have only used heimdall.
I will retry with this serie when possible.

File: cmd/pxe.c [hunk 1]
```diff
@@ -1709,7 +1759,11 @@ static int do_sysboot(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 	if (prompt)
 		cfg->prompt = 1;
 
+#ifdef CONFIG_BOOTMENU
```
Comment (PabloPL) [inline cmd/pxe.c]: Shouldn't here be #ifndef ? Because currently it looks like handle_pxe_menu wont' be called for devices without BOOTMENU

Comment (xc-racer99) [reply cmd/pxe.c]: Thanks, good catch.  Should be #ifndef CONFIG_CMD_BOOTMENU - I had the wrong define as well.

Comment (xc-racer99) [reply cmd/pxe.c]: Force-pushed with correcting the ifdef, also ifdefing to use the bootmenu when possible when doing true PXE boot.

**Review by PabloPL — APPROVED** [review]

**Comment by PabloPL** [comment]
Tested on i9000 with pmOS (port done by @xc-racer99) flashed on mmc and it works fine.

**Comment by kwizart** [comment]
Okay, I've managed to boot a fedora 30 boot menu where I can choose the kernel.
So the fedora 30 partition layout is detected as appropriate. 

Next step would be to support uEFI on armv7 !
See also https://fedoraproject.org/wiki/Changes/uEFIforARMv7

I need to setup a kernel that has the needed soc/board changes added

I have a small issue that:
- When attempting to use a given boot menu, I've selected a wrong option (boot into mmc instead of sdcard).
that left me with a u-boot prompt and I was not able to shutdown the device (since power is enter, there).
I had to remove the battery and reflash u-boot because the old menu choice was saved...
- There is a little screen glitch that wasn't available previously IIRC, this is minor.

So far, good work !

**Comment by PabloPL** [comment]
@kwizart 
> When attempting to use a given boot menu, I've selected a wrong option (boot into mmc instead of sdcard).
that left me with a u-boot prompt and I was not able to shutdown the device (since power is enter, there).

If during boot You hold home key (on i9000), it will always boot into bootmenu (no matter what was last booted) - no need to reflash uboot :)

I think we could/should update documentation - for example wiki here

Edit 1:
Added information about boot modes into https://github.com/xc-racer99/u-boot-galaxys4g/wiki
in section named "BOOT modes"

**Comment by xc-racer99** [comment]
>     * There is a little screen glitch that wasn't available previously IIRC, this is minor.

That's because we're now using the u-boot Exynos FB driver as opposed to re-using the SBL.

@PabloPL Thanks for adding the notes to the wiki about which buttons to use.

@kwizart You mentioned a few months ago about not being able to get into the bootmenu.  Is this still the case or now that you know the previously undocumented secret it works properly :)

Since it looks as if this is working for everybody, I'm going to go ahead and merge it.

**Pull request merged**